V2 - Tradução: datetimes.qmd (#109)
cienciadedatos/pt-r4ds: datetimes.qmd

```{r}
#| message: false
library(tidyverse)
library(dados)
```

## Criando datas/horários {#sec-creating-datetimes}

Existem três tipos de dados de data/horário que se referem a um instante no tempo:

-   Uma **data**.
    Tibbles imprimem isso como `<date>`.

-   Um **horário** dentro de um dia.
    Tibbles imprimem isso como `<time>`.

> -   Uma **data-horário** é uma data mais um horário: identifica exclusivamente um instante no tempo (tipicamente até o segundo mais próximo).

Author: ```suggestion
-   Uma **data-hora** é uma data mais um horário: identifica exclusivamente um instante no tempo (tipicamente até o segundo mais próximo).
```

    Tibbles imprimem isso como `<dttm>`.
    O R base chama isso de POSIXct, mas não é exatamente fácil de pronunciar.

> Neste capítulo iremos nos concentrar nas classes data (*date*) e data-horário (*date-time*), já que o R não possui uma classe nativa para lidar apenas com horário (*time*).

Author: ```suggestion
Neste capítulo iremos nos concentrar nas classes data (*date*) e data-hora (*date-time*), já que o R não possui uma classe nativa para lidar apenas com horário (*time*).
```

Se você precisar de uma, pode usar o pacote **hms**.

Você deve sempre usar o tipo de dado mais simples possível que funcione para suas necessidades.
Isso significa que, se você puder usar uma data em vez de uma data-horário, você deve.

> Datas-horário são substancialmente mais complicadas devido à necessidade de lidar com fusos horários, o que abordaremos no final do capítulo.

Author: ```suggestion
Datas-hora são substancialmente mais complicadas devido à necessidade de lidar com fusos horários, o que abordaremos no final do capítulo.
```

> Para obter a data ou data-horário atual, você pode usar `today()` ou `now()`:

Author: ```suggestion
Para obter a data ou data-hora atual, você pode usar `today()` ou `now()`:
```

```{r}
today()
now()
```

Caso contrário, as seções a seguir descrevem as quatro maneiras mais prováveis de criar uma data/hora:

> -   Durante a leitura de um arquivo com readr.

Author: ```suggestion
-   Durante a leitura de um arquivo com o pacote readr.
```

> -   A partir de uma string.

Author: ```suggestion
-   A partir de um texto (*string*).
```

dmy("31-Jan-2017")
```

`ymd()` e funções similares criam datas.

> Para criar uma data-hora, adicione um sublinhado e uma ou mais de "h", "m" e "s" ao nome da função de análise:

Author: ```suggestion
Para criar uma data-hora, adicione um *underscore* (`_`) e uma ou mais de "h", "m" e "s" ao nome da função de análise (*parsing*):
```

dmy("31-Jan-2017")
```

> `ymd()` e amigos criam datas.

Author: ```suggestion
`ymd()` e funções similares criam datas.
```

> Para criar uma data-horário, adicione um sublinhado e uma ou mais de "h", "m" e "s" ao nome da função de análise:

Author: ```suggestion
Para criar uma data-hora, adicione um sublinhado e uma ou mais de "h", "m" e "s" ao nome da função de análise:
```

> Para criar uma data-horário, adicione um sublinhado e uma ou mais de "h", "m" e "s" ao nome da função de análise:

Author: Esse *sublinhado* tá esquisito. Ver o contexto e pensar no que encaixa

```{r}
ymd_hms("2017-01-31 20:11:59")
mdy_hm("01/31/2017 08:01")
```

> Você também pode forçar a criação de uma data-horário a partir de uma data fornecendo um fuso horário:

Author: ```suggestion
Você também pode forçar a criação de uma data-hora a partir de uma data fornecendo um fuso horário:
```

```{r}
ymd("2017-01-31", tz = "UTC")
```

> Aqui eu uso o fuso horário UTC[^datetimes-3] que você também pode conhecer como GMT, ou Greenwich Mean Time, o horário em 0° de longitude[^datetimes-4]

Author: ```suggestion
Aqui eu uso o fuso horário UTC[^datetimes-3] que você também pode conhecer como GMT, ou Greenwich Mean Time, o horário em 0° de longitude[^datetimes-4].
```

> . Ele não usa o horário de verão, o que torna os cálculos um pouco mais simples
> .

Author: ```suggestion
Ele não usa o horário de verão, o que torna os cálculos um pouco mais simples.
```

[^datetimes-3]: Você pode se perguntar o que significa UTC.

>     É um compromisso entre o inglês "Coordinated Universal Time" e o francês "Temps Universel Coordonné".

Author: ```suggestion
    É um meio-termo entre o inglês "Coordinated Universal Time" e o francês "Temps Universel Coordonné".
```
Esse "compromise" tá estranho. não entendi bem o que os autores quiseram dizer.

Author: Parei aqui

Reviewer: Provavelmente combinação

Author: Achei na internet a tradução "meio-termo" e acho que faz sentido!

[^datetimes-4]: Sem prêmios para quem adivinhar qual país criou o sistema de longitude.

### A partir de componentes individuais

Em vez de uma única string, às vezes você terá os componentes individuais da data-hora espalhados por várias colunas.
Isso é o que temos nos dados `voos`:

```{r}
voos |> 
  select(ano, mes, dia, hora, minuto)
```

> Para criar uma data/horário a partir deste tipo de entrada, use a função `make_date()` para datas, ou `make_datetime()` para data-horário:

Author: ```suggestion
Para criar uma data/horário a partir deste tipo de entrada, use a função `make_date()` para datas, ou `make_datetime()` para data-hora:
```

```{r}
make_datetime_100 <- function(ano, mes, dia, horario) {
  make_datetime(ano, mes, dia, horario %/% 100, horario %% 100)
}

voos_dt <- voos |> 
  filter(!is.na(horario_saida), !is.na(horario_chegada)) |> 
  mutate(
    horario_saida = make_datetime_100(ano, mes, dia, horario_saida),
    horario_chegada = make_datetime_100(ano, mes, dia, horario_chegada),
    saida_programada = make_datetime_100(ano, mes, dia, saida_programada),
    chegada_prevista = make_datetime_100(ano, mes, dia, chegada_prevista)
  ) |> 
  select(origem, destino, starts_with("atraso"), 
         horario_saida, saida_programada,
         horario_chegada, chegada_prevista,
         tempo_voo)
```

Com esses dados, podemos visualizar a distribuição dos horários de partida ao longo do ano:

```{r}
#| fig.alt: >
#|   Um gráfico de polígono de frequência com horário de saída (janeiro-dezembro de 2013) no eixo-x
#|   e número de voos no eixo-y (0-1000). O polígono de frequência
#|   é agrupado por dia para que você veja uma série temporal de voos por dia. O
#|   o padrão é dominado por um padrão semanal; há menos voos 

> #|   nos fins de semana. São poucos os dias que se destinoacam por terem tido surpreendentemente

Author: ```suggestion
#|   nos fins de semana. São poucos os dias que se destacam por terem tido surpreendentemente
```

  geom_freqpoly(binwidth = 600) # 600 s = 10 minutos
```

> Observe que quando você usa datas-hora em um contexto numérico (como em um histograma), 1 significa 1 segundo, então uma largura de bin de 86400 significa um dia.

Author: ```suggestion
Observe que quando você usa datas-hora em um contexto numérico (como em um histograma), 1 significa 1 segundo, então uma largura de intervalo de classe (*bin*) de 86400 significa um dia.
```

  geom_freqpoly(binwidth = 600) # 600 s = 10 minutos
```

> Observe que quando você usa datas-horários em um contexto numérico (como em um histograma), 1 significa 1 segundo, então uma largura de bin de 86400 significa um dia.

Author: ```suggestion
Observe que quando você usa datas-hora em um contexto numérico (como em um histograma), 1 significa 1 segundo, então uma largura de bin de 86400 significa um dia.
```

> Observe que quando você usa datas-horários em um contexto numérico (como em um histograma), 1 significa 1 segundo, então uma largura de bin de 86400 significa um dia.

Author: ```suggestion
Observe que quando você usa datas-horários em um contexto numérico (como em um histograma), 1 significa 1 segundo, então uma largura de uma barra (bin) de 86400 significa um dia.
```

> Observe que quando você usa datas-horários em um contexto numérico (como em um histograma), 1 significa 1 segundo, então uma largura de bin de 86400 significa um dia.

Author: @cesar-galvao nesse caso, bin seria barra ou caixa? ou outra forma?

  mutate(minuto = minute(horario_saida)) |> 
  group_by(minuto) |> 
  summarize(
    media_atraso = mean(atraso_saida, na.rm = TRUE),
    n = n()
  ) |> 
  ggplot(aes(x = minuto, y = media_atraso)) +
  geom_line()
```

Curiosamente, se olharmos para o horário de saída *programado*, não veremos um padrão tão forte:

```{r}
#| fig-alt: | 
#|   Um gráfico de linhas com o minuto da partida programada (0-60) no eixo x

> #|   e atraso médio (4-16). Há relativamente pouco padrão, apenas uma

Author: "Há relativamente pouco padrão," soa esquisito. Alguma sugestão?

Reviewer: Talvez "pouca tendência evidente"?

Author: ```suggestion
#|   e atraso médio (4-16). Há pouca tendência evidente, apenas uma
```
talvez assim fica melhor? @cesar-galvao

Reviewer: @beatrizmilz acho que a ideia é a mesma de uma distribuição uniforme. O padrão é pouco discernível. Acho que 'pouco padrao' nao é tão simples de entender pelo leitor comum. Penso que 'pouco tendencia evidente' deixe mais claro.

Author: beleza! vou usar sua sugestão!

#|   pequena sugestão de que o atraso médio diminui de talvez 10 minutos
#|   para 8 minutos ao longo da hora.
chegada_prevista <- voos_dt |> 
  mutate(minuto = minute(saida_programada)) |> 
  group_by(minuto) |> 
  summarize(
    media_atraso = mean(atraso_chegada, na.rm = TRUE),
    n = n()
  )

ggplot(chegada_prevista, aes(x = minuto, y = media_atraso)) +
  geom_line()
```

Então, por que vemos esse padrão com os horários reais de partida?
Bem, como muitos dados coletados por humanos, há um forte viés em relação aos voos que partem em horários de partida "agradáveis", como mostra @fig-human-rounding.

> Esteja sempre atento a esse tipo de padrão sempre que trabalhar com dados que envolvem julgamento humano!

Author: Parei aqui!

datetime
```

> Alternativamente, em vez de modificar uma variável existente, você pode criar uma nova data-horário com `update()`.

Author: ```suggestion
Alternativamente, em vez de modificar uma variável existente, você pode criar uma nova data-hora com `update()`.
```

Como sempre, escolha a estrutura de dados mais simples que resolva seu problema.
Se você se preocupa apenas com o tempo físico, use uma duração; se você precisar adicionar períodos de tempo, use um período; se você precisar descobrir quanto tempo dura um intervalo entre unidades, use um intervalo.

### Durações

No R, ao subtrair duas datas, você obtém um objeto de intervalo de tempo da classe *difftime*:

```{r}

> # Quantos anos tem Hadley?

Author: ```suggestion
# Quantos anos o Hadley tem?
```

Sys.timezone()
```

> (Se o R não souber, você receberá um `NA`.) E veja a lista completa de todos os nomes de fusos horários com `OlsonNames()`:

Author: ```suggestion
(Se o R não souber, você receberá um `NA`.)

Você também pode ver a lista completa de todos os nomes de fusos horários com `OlsonNames()`:
```

UTC (Tempo Universal Coordenado) é o fuso horário padrão usado pela comunidade científica e é aproximadamente equivalente ao GMT (Horário de Greenwich).
Ele não possui horário de verão (DST), o que o torna uma representação conveniente para cálculos.
Operações que combinam datas-horas, como `c()`, frequentemente descartam o fuso horário.
Nesse caso, as datas-horas serão exibidas no fuso horário do primeiro elemento:

```{r}
x4 <- c(x1, x2, x3)
x4
```

Você pode alterar o fuso horário de duas maneiras:

-   Manter o instante no tempo o mesmo e alterar como ele é exibido.
    Use isso quando o instante estiver correto, mas você quiser uma exibição mais natural.

    ```{r}
    x4a <- with_tz(x4, tzone = "Australia/Lord_Howe")
    x4a
    x4a - x4
    ```

    (Isso também ilustra outro desafio dos fusos horários: nem todos são deslocamentos de horas inteiras!)

    -   Alterar o instante subjacente no tempo. Use isso quando você tiver um instante que foi rotulado com o fuso horário incorreto e precisar corrigi-lo.

    ```{r}
    x4b <- force_tz(x4, tzone = "Australia/Lord_Howe")
    x4b
    x4b - x4
    ```

## Resumo

> Este capítulo apresentou as ferramentas que o lubridate fornece para ajudá-lo a trabalhar com dados de data e hora.

Author: ```suggestion
Este capítulo apresentou as ferramentas que o lubridate fornece para te ajudar a trabalhar com dados de data e hora.
```

> Trabalhar com datas e horários pode parecer mais difícil do que o necessário, mas espero que este capítulo tenha ajudado você a entender o porquê --- datas e horários são mais complexos do que parecem à primeira vista, e lidar com todas as situações possíveis adiciona complexidade.

Author: ```suggestion
Trabalhar com datas e horários pode parecer mais difícil do que o necessário, mas espero que este capítulo tenha ajudado você a entender o porquê --- datas e horários são mais complexas do que parecem à primeira vista, e lidar com todas as situações possíveis adiciona complexidade.
```

Mesmo que seus dados nunca cruzem um limite de horário de verão ou envolvam um ano bissexto, as funções precisam ser capazes de lidar com isso.

> O próximo capítulo faz um resumo dos valores ausentes.

Author: ```suggestion
O próximo capítulo faz um resumo dos valores ausentes (`NA`).
```
